Std/Classes/RatCast.lean

```diff
@@ -0,0 +1,23 @@
+/-
+Copyright (c) 2014 Robert Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert Lewis, Leonardo de Moura, Johannes Hölzl, Mario Carneiro, Gabriel Ebner
+-/
+import Std.Data.Rat.Basic
+
+/-- Type class for the canonical homomorphism `Rat → K`. -/
+class RatCast (K : Type u) where
+  /-- The canonical homomorphism `Rat → K`. -/
+  protected ratCast : Rat → K
+
+/-- Canonical homomorphism from `Rat` to a division ring `K`.
+This is just the bare function in order to aid in creating instances of `DivisionRing`. -/
+@[coe, match_pattern] protected def Rat.cast {K : Type u} [RatCast K] : Rat → K := RatCast.ratCast
+
+-- see note [coercion into rings]
+instance [RatCast K] : CoeTail Rat K where coe := Rat.cast
+
+-- see note [coercion into rings]
+instance [RatCast K] : CoeHTCT Rat K where coe := Rat.cast
+
+instance [RatCast α] : OfScientific α where ofScientific m s e := (Rat.ofScientific m s e : α)
```

Comment:
```suggestion
instance [DivisionRing α] : OfScientific α where ofScientific m s e := (Rat.ofScientific m s e : α)
```
I thought this was the plan?

Reply:
I think it was (is?). The trouble with this is that it leaves std without an `OfScientific Rat` / `RatCast Rat` instance since it doesn't have `DivisionRing`.

Reply:
It was indeed the plan; however, `DivisionRing` requires mathlib4 (`Mathlib.Algebra.Field.Defs`), so we'd have to either move `DivisionRing` into std or move this instance out of std.

I will say that I've just updated leanprover-community/mathlib4#1707 locally with these std changes to see if everything worked with this new instance, and it does. :)

On reflection, using `RatCast` for this instance also seems reasonable to me: if you know how to turn a rational number into an element of your type, you know how to turn a decimal expression into an element of your type. This new instance effectively says "decimals are simply rational numbers", even if you don't respect the algebraic properties of the rationals.

Reply:
> This new instance effectively says "decimals are simply rational numbers"

The question when making this an instance is not whether it's a reasonable way.  But whether it's the *only* reasonable way.

For example, we probably want a `RatCast` instance for `Float`.  But without changing core, this will give rise to a different `OfScientific` instance for `Float`.  And then you end up in a situation where you have two `-0` literals that are not equal, which is something I'd really like to avoid.  (Note that `(0 : Float)` and `(-0 : Float)` (should) evaluate to two different numbers.)

This was the reason why I was okay to have the `[DivisionRing R] : OfScientific R` instance.  Because at least then there's only one reasonable choice for the instance.  (And the risk for non-defeq diamonds is lower since we're excluding floats.)

Reply:
> I think it is possible to do this with overlapping instances (e.g. `OfScientific Rat` here and `[DivisionRing α] : OfScientific α` in mathlib), as long as we can make them defeq at instance reducibility (which is what is needed for `norm_num` to work).

I haven't had much luck doing this myself. In a file where I can evaluate it, I get
```
attribute [-instance] instOfScientific

instance y : OfScientific Rat where ofScientific := Rat.ofScientific
instance x [DivisionRing α] : OfScientific α where ofScientific m s e := (y.ofScientific m s e : α)
instance x' [DivisionRing α] : OfScientific α where ofScientific m s e := (Rat.ofScientific m s e : α)

example : y = x := by with_reducible_and_instances rfl -- fails
example : y = x' := by with_reducible_and_instances rfl -- fails
```
Is there a better way?

Reply:
I think it would be okay to keep the instance here as an `abbrev` and only use it to define an instance for division rings in mathlib4.  (And maybe for rationals already in std4.)

Reply:
> I think it is possible to do this with overlapping instances (e.g. `OfScientific Rat` here and `[DivisionRing α] : OfScientific α` in mathlib), as long as we can make them defeq at instance reducibility (which is what is needed for `norm_num` to work).

What would that achieve?  The point here is that there shouldn't be a generic `[RatCast R] : OfScientific R` instance, because it causes diamonds with floats.  Adding an extra instance doesn't help with that.

Reply:
> The point here is that there shouldn't be a generic `[RatCast R] : OfScientific R` instance, because it causes diamonds with floats.

Does it? `#synth RatCast Float` fails for me. Is this expected to change in the future?

Reply:
Also I think Mario is proposing using `OfScientific Rat` and `[DivisionRing α] : OfScientific α` instead of `[RatCast α] : OfScientific α`, not in addition to it. But we would still need to solve the defeq problem somehow (assuming I'm testing it correctly).

Reply:
Would giving the `[RatCast α] : OfScientific α` instance a lower priority work? Or would that be too fragile even if it did work?

Reply:
It's not an issue of priority. You can get diamonds due to composing lemmas which each made the best available choice for the context in which they were stated.

Reply:
I figured out a way to make the instances of `OfScientific Rat` and `[DivisionRing α] : OfScientific α` when `α` is `Rat` defeq: use `Rat.cast` in both.
```
instance : RatCast Rat where ratCast := id

instance x : OfScientific Rat where ofScientific m s e := Rat.cast (Rat.ofScientific m s e)

/-- mock division ring -/
class DivisionRing (α) extends RatCast α

instance : DivisionRing Rat where ratCast := id

instance y [DivisionRing α] : OfScientific α where
  ofScientific m s e := (Rat.ofScientific m s e : α)

example : x = y := by with_reducible_and_instances rfl -- works
```

Reply:
There is no need for that, you can just avoid `id` in the `DivisionRing Rat` instance and it will also work:
```lean
instance x : OfScientific Rat where ofScientific m s e := Rat.ofScientific m s e
...
instance : DivisionRing Rat where ratCast x := x
...
example : x = y := by with_reducible_and_instances rfl -- works
```